server/base/src/main/java/org/apache/accumulo/server/conf/codec/VersionedProperties.java

```diff
@@ -0,0 +1,258 @@
+   * <p>
+   * It is much more efficient to add multiple properties at a time rather than one by one.
+   * <p>
+   * Because instances of this class are immutable, this method creates a new copy of the
+   * properties. Other processes will continue to see original values retrieved from the data store.
+   * Other processes will receive an update when the instance is encoded and stored in the data
+   * store and then retrieved with the normal store update mechanisms.
+   *
+   * @param key
+   *          the property name.
+   * @param value
+   *          the property value.
+   * @return A new instance of this class with the property added or updated.
+   */
+  public VersionedProperties addOrUpdate(final String key, final String value) {
+    ImmutableMap<String,String> updated =
+        ImmutableMap.<String,String>builder().putAll(new HashMap<>() {
+          {
+            putAll(props);
+            put(key, value);
+          }
+        }).build();
+    return new VersionedProperties(dataVersion, Instant.now(), updated);
+  }
+
+  /**
+   * Add or update multiple properties. If a property already exists it is overwritten.
+   * <p>
+   * Because instances of this class are immutable, this method creates a new copy of the
+   * properties. Other processes will continue to see original values retrieved from the data store.
+   * Other processes will receive an update when the instance is encoded and stored in the data
+   * store and then retrieved with the normal store update mechanisms.
+   *
+   * @param updates
+   *          A map of key, values pairs.
+   * @return A new instance of this class with the properties added or updated.
+   */
+  public VersionedProperties addOrUpdate(final Map<String,String> updates) {
+    ImmutableMap<String,String> updated =
+        ImmutableMap.<String,String>builder().putAll(new HashMap<>() {
+          {
+            putAll(props);
+            putAll(updates);
+          }
+        }).build();
+    return new VersionedProperties(dataVersion, Instant.now(), updated);
+  }
+
+  /**
+   * Delete multiple properties provided as a collection of keys.
+   * <p>
+   * Because instances of this class are immutable, this method creates a new copy of the
+   * properties. Other processes will continue to see original values retrieved from the data store.
+   * Other processes will receive an update when the instance is encoded and stored in the data
+   * store and then retrieved with the normal store update mechanisms.
+   *
+   * @param keys
+   *          a collection of the keys that if they exist, will be removed.
+   * @return A new instance of this class.
+   */
+  public VersionedProperties remove(Collection<String> keys) {
+
+    HashMap<String,String> orig = new HashMap<>(props);
+    keys.forEach(orig::remove);
+
+    ImmutableMap<String,String> updated =
+        ImmutableMap.<String,String>builder().putAll(orig).build();
+
```

Comment:
Avoid unnecessary complexity, since the constructor will make an immutable copy anyway and this is easier to understand and maintain:
```suggestion
    var updated = new HashMap<>(props);
    updated.keySet().removeAll(keys); // removing from keySet view removes from underlying mapping
```